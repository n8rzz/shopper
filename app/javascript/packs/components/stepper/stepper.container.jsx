import React from 'react';
import PropTypes from 'prop-types';

import ApiService from '../../service/api.service';
import EventService from '../../service/event.service';
import { Stepper } from './stepper';
import { EVENT_NAME } from '../../constants/event-names';

const CREATE_ORDER_ITEM_URL = '/order_items/create/item.json';

export class StepperContainer extends React.Component {
    constructor(props) {
        super(props);

        this.onChangeQtyHandler = this._onChangeQty.bind(this);
        this.onClickDecreaseHandler = this._onClickDecrease.bind(this);
        this.onClickIncreaseHandler = this._onClickIncrease.bind(this);
        this.onClickSubmitHandler = this._onClickSubmit.bind(this);
        this.onSubmitSuccessHandler = this._onSubmitSuccess.bind(this);

        this.state = {
            qty: 1,
            isSubmitting: false,
            isSubmitSuccess: false,
        };
    }

    componentWillUnmount() {
        this.onChangeQtyHandler = null;
        this.onClickDecreaseHandler = null;
        this.onClickIncreaseHandler = null;
        this.onClickSubmitHandler = null;
        this.onSubmitSuccessHandler = null;
    }

    _onChangeQty(event) {
        if (event.currentTarget.value === '') {
            return;
        }

        this.setState({ qty: event.currentTarget.value });
    }

    _onClickDecrease() {
        // eslint-disable-next-line react/no-access-state-in-setstate
        let nextQty = this.state.qty - 1;

        if (nextQty < 1) {
            nextQty = 1;
        }

        this.setState({ qty: nextQty });
    }

    _onClickIncrease() {
        this.setState((prevState) => ({ qty: prevState.qty + 1 }));
    }

    _onClickSubmit(event) {
        event.stopPropagation();
        event.preventDefault();

        if (this.state.qty < 1) {
            return;
        }

        const orderItem = {
            item_id: this.props.itemId,
            department_id: this.props.departmentId,
            qty: this.state.qty,
        };

        this.setState({ isSubmitting: true }, this._callSubmit(orderItem));
    }

    _callSubmit(orderItem) {
        ApiService.post(CREATE_ORDER_ITEM_URL, orderItem, this.props.csrf)
            .then((response) => {
                if (response.status >= 300) {
                    console.error(`
                        Received an unexpected status code from ${CREATE_ORDER_ITEM_URL}. See response: ${response}
                    `);

                    return;
                }
                // added to pending order`);
                const notice = `${this.props.itemName} added to order`;

                EventService.emit(EVENT_NAME.NOTICE_SUCCESS, notice);

                const nextState = {
                    qty: 1,
                    isSubmitting: false,
                    isSubmitSuccess: true,
                };
                this.setState(nextState, this.onSubmitSuccessHandler);
            })
            .catch((error) => {
                // TODO: add ability to set flash warning, error
                // this._flashController.showNoticeMessage('Something went wrong, Item
                // was not successfully added to Order');

                throw error;
            });
    }

    _onSubmitSuccess() {
        setTimeout(() => {
            this.setState({ isSubmitSuccess: false });
        }, 3000);
    }

    render() {
        return (
            <Stepper
                itemId={this.props.itemId}
                qty={this.state.qty}
                isSubmitting={this.state.isSubmitting}
                isSubmitSuccess={this.state.isSubmitSuccess}
                onChangeQtyHandler={this.onChangeQtyHandler}
                onClickDecreaseHandler={this.onClickDecreaseHandler}
                onClickIncreaseHandler={this.onClickIncreaseHandler}
                onClickSubmitHandler={this.onClickSubmitHandler}
            />
        );
    }
}

StepperContainer.propTypes = {
    csrf: PropTypes.string.isRequired,
    departmentId: PropTypes.number.isRequired,
    itemId: PropTypes.number.isRequired,
    itemName: PropTypes.string.isRequired,
};
