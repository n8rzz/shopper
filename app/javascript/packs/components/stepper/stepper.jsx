import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

export class Stepper extends React.Component {
    buildClassnames() {
        return classnames({
            'stepper-ft-submit': true,
            'mix-btn_isLoading': this.props.isSubmitting,
            'mix-btn_isSuccess': this.props.isSubmitSuccess,
        });
    }

    _buildStepperBody() {
        return (
            <div className={'stepper-bd'}>
                <div
                    className={'stepper-bd-btn js-stepper-btn-decrease'}
                    onClick={this.props.onClickDecreaseHandler}
                >
                    -
                </div>
                <div id={`item-${this.props.itemId}-qty`}>
                    {this.props.qty}
                </div>
                <div
                    className={'stepper-bd-btn js-stepper-btn-increase'}
                    onClick={this.props.onClickIncreaseHandler}
                >
                    +
                </div>
            </div>
        );
    }

    _buildSubmitButton() {
        let buttonValue = this.props.isSubmitting
            ? 'Adding...'
            : 'Add to order';

        if (this.props.isSubmitSuccess) {
            buttonValue = 'Success!';
        }

        return (
            <div className={'stepper-ft'}>
                <input
                    type={'submit'}
                    value={buttonValue}
                    className={this.buildClassnames()}
                />
            </div>
        );
    }

    render() {
        return (
            <form onSubmit={this.props.onClickSubmitHandler}>
                <div
                    className={'stepper'}
                    data-stepper-id={this.props.itemId}
                >
                    { this._buildStepperBody() }
                    { this._buildSubmitButton() }
                </div>
            </form>
        );
    }
}

Stepper.propTypes = {
    itemId: PropTypes.number.isRequired,
    qty: PropTypes.number.isRequired,
    isSubmitting: PropTypes.bool.isRequired,
    isSubmitSuccess: PropTypes.bool.isRequired,
    onClickDecreaseHandler: PropTypes.func.isRequired,
    onClickIncreaseHandler: PropTypes.func.isRequired,
    onClickSubmitHandler: PropTypes.func.isRequired,
};
