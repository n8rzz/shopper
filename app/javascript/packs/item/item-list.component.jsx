import React from 'react';
import PropTypes from 'prop-types';
import { Item } from './item.component';

export function ItemList(props) {
    if (props.itemList.length === 0) {
        return null;
    }

    return (
        <ul className={'vlist'}>
            {props.itemList.map((item, index) => (
                <Item
                    csrf={props.csrf}
                    departmentId={item.department_id}
                    departmentName={props.departmentMap[item.department_id].name}
                    itemId={item.id}
                    itemName={item.name}
                    key={`item-${index}`}
                />
            ))}
        </ul>
    );
}

ItemList.propTypes = {
    csrf: PropTypes.string.isRequired,
    departmentMap: PropTypes.object.isRequired,
    itemList: PropTypes.array.isRequired,
};
