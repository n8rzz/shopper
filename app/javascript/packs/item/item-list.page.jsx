import React from 'react';
import PropTypes from 'prop-types';
import _groupBy from 'lodash/groupBy';
import _sortBy from 'lodash/sortBy';
import {FilterBar} from '../components/filter-bar/filter-bar.component';
import {ItemList} from './item-list.component';
import { GroupedItemList } from './grouped-item-list.component';

// TODO: abstract these constants to another file
export const FILTER_CONCERN = {
    ALPHA: 'ALPHA',
    DEPARTMENT: 'DEPARTMENT',
    FILTER: 'FILTER',
};
export const FILTER_LABEL = {
    [FILTER_CONCERN.ALPHA]: 'Alpha',
    [FILTER_CONCERN.DEPARTMENT]: 'Department',
    [FILTER_CONCERN.FILTER]: 'Filter',
}
export const FILTER_BAR_ITEMS = [
    {
        text: FILTER_LABEL.ALPHA,
        url: '',
    },
    {
        text: FILTER_LABEL.DEPARTMENT,
        url: '',
    },
    {
        text: FILTER_LABEL.FILTER,
        url: '',
    },
];
export const DEFAULT_DRILL_DOWN_VALUE = '--';

export class ItemListPage extends React.Component {
    constructor(props) {
        super(props);

        this._departmentMap = this._buildDepartmentMap(props);
        this._itemsByFirstLetter = this._buildItemsByFirstLetter(props);
        this._filterableItemList = Object.keys(this._itemsByFirstLetter);
        this.state = {
            concern: FILTER_CONCERN.ALPHA,
            itemsByFilter: props.items,
        };
    }

    componentWillMount() {
        this._onChangeFilterHandler = this._onChangeFilter.bind(this);
        this._onChangeLetterFilterValueHandler = this._onChangeLetterFilterValue.bind(this);
    }

    componentWillUnmount() {
        this._onChangeFilterHandler = null;
        this._onChangeLetterFilterValueHandler = null;
    }

    _buildSecondaryFilterOptionListJsx() {
        return (
            <select
                className="input-select"
                onChange={this._onChangeLetterFilterValueHandler}
            >
                <option value={DEFAULT_DRILL_DOWN_VALUE}>{DEFAULT_DRILL_DOWN_VALUE}</option>
                {this._filterableItemList.map((filterableLetter, index) => {
                    return (
                        <option value={filterableLetter} key={`select-${index}`}>
                            {filterableLetter.toUpperCase()}
                        </option>
                    );
                })}
            </select>
        );
    }

    _buildItemListForActiveConcernJsx() {
        switch (this.state.concern) {
            case FILTER_CONCERN.DEPARTMENT:
                return (
                    <GroupedItemList
                        csrf={this.props.csrf}
                        departments={this.props.departments}
                        departmentMap={this._departmentMap}
                        items={this.props.items}
                    />
                );
            case FILTER_CONCERN.FILTER:
                return (
                    <React.Fragment>
                        <div className="vr_3">
                            {this._buildSecondaryFilterOptionListJsx()}
                        </div>

                        <ItemList
                            csrf={this.props.csrf}
                            departmentMap={this._departmentMap}
                            itemList={this.state.itemsByFilter}
                        />
                    </React.Fragment>
                );
            default:
                return (
                    <ItemList
                        csrf={this.props.csrf}
                        departmentMap={this._departmentMap}
                        itemList={this.props.items}
                    />
                );
        }
    }

    render() {
        return (
            <React.Fragment>
                <FilterBar
                    items={FILTER_BAR_ITEMS}
                    onChangeFilterHandler={this._onChangeFilterHandler}
                />

                {this._buildItemListForActiveConcernJsx()}
            </React.Fragment>
        );
    }

    // TODO: rename this
    _onChangeLetterFilterValue(event) {
        const nextLetterFilter = event.target.value;
        let itemsByFilter = this._itemsByFirstLetter[nextLetterFilter];

        if (nextLetterFilter === DEFAULT_DRILL_DOWN_VALUE) {
            itemsByFilter = this.props.items;
        }

        this.setState({ itemsByFilter });
    }

    _onChangeFilter(event) {
        event.preventDefault();

        const filterName = event.currentTarget.text.toUpperCase();

        this.setState({ concern: FILTER_CONCERN[filterName] });
    }

    _buildDepartmentMap(props) {
        return props.departments.reduce((sum, department) => {
            sum[department.id] = department;

            return sum;
        }, {});
    }

    _buildItemsByFirstLetter(props) {
        return props.items.reduce((sum, item) => {
            const firstLetter = item.name.substr(0, 1).toLowerCase();

            if (typeof sum[firstLetter] !== 'undefined') {
                sum[firstLetter] = [
                    ...sum[firstLetter],
                    item,
                ];

                return sum;
            }

            sum[firstLetter] = [item];

            return sum;
        }, {});
    }
}

ItemListPage.propTypes = {
    csrf: PropTypes.string.isRequired,
    departments: PropTypes.array,
    items: PropTypes.array,
};
