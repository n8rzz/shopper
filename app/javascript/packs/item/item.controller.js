import FlashController from '../flash/flash.controller';
import OrderItemModel from './order-item.model';
import ApiService from '../service/api.service';

const CLASSNAMES = {
    ROOT_ELEMENT: 'js-item',
    ACTION_ITEM: 'js-item-action',
    ADD_ITEM_TO_ORDER_FORM: 'js-item-createOrderItemForm'
};

export class OrderController {
    constructor() {
        const $element = document.getElementsByClassName(CLASSNAMES.ROOT_ELEMENT);

        if ($element.length === 0) {
            this.disable();

            return;
        }

        this._$actionItemElementList = null;
        this._flashController = new FlashController();
        this._isEnabled = false;

        return this._init()
            ._createChildren()
            ._setupHandlers()
            .enable();
    }

    _init() {
        return this;
    }

    _createChildren() {
        this._$actionItemElementList = document.getElementsByClassName(CLASSNAMES.ACTION_ITEM);

        return this;
    }

    _setupHandlers() {
        this._onClickAddItemToOrderHandler = this._onClickAddItemToOrder.bind(this);

        return this;
    }

    enable() {
        if (this._isEnabled) {
            return this;
        }

        this._addHandlerToElementList(this._$actionItemElementList, this._onClickAddItemToOrderHandler);

        this._isEnabled = true;

        return this;
    }

    disable() {
        if (!this._isEnabled) {
            return this;
        }

        this._isEnabled = false;

        return this;
    }

    _addHandlerToElementList(elementList, handler) {
        if (elementList.length === 0) {
            return;
        }

        for (let i =0; i < elementList.length; i++) {
            const element = elementList[i];

            element.addEventListener('click', handler);
        }
    }

    _onClickAddItemToOrder(event) {
        event.stopPropagation();
        event.preventDefault();

        const csrf = event.currentTarget.dataset.csrf;
        const createOrderItemUrl = `/order_items/create/item.json`;
        const orderItemModel = new OrderItemModel(event.currentTarget);

        if (!orderItemModel.isValid()) {
            return;
        }

        ApiService.post(createOrderItemUrl, orderItemModel.toJson(), csrf)
            .then((response) => {
                if (response.status >= 300) {
                    console.error(`Received an unexpected status code from ${createOrderItemUrl}. See response: ${response}`);

                    return;
                }

                this._flashController.showNoticeMessage(`${orderItemModel.name} added to pending order`);
            })
            .catch((error) => {
                // TODO: add ability to set flash warning, error
                this._flashController.showNoticeMessage('Something went wrong, Item was not successfully added to Order');

                throw error;
            });
    }
}

document.addEventListener('turbolinks:load', () => {
    const singleOrderController = new OrderController();
});
