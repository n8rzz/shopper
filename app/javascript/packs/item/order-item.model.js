/**
 * Accepts a DOM element (`currentTarget` from a click event) and
 * extracts information from that DOM element used in submission
 * of a form.
 *
 * This model acts as a flyweight and should only be available
 * between mouse click and api request.
 *
 * @class OrderItemModel
 */
export default class OrderItemModel {
    constructor(element) {
        this.itemId = -1;
        this.name = '';
        this.departmentId = -1;
        this.qty = -1;

        return this._init(element);
    }

    isValid() {
        return this.itemId !== -1
            && this.departmentId !== -1
            && this.qty !== -1
            && this.qty > 0;
    }

    toJson() {
        return {
            item_id: this.itemId,
            department_id: this.departmentId,
            qty: this.qty,
        };
    }

    _init(element) {
        this.itemId = parseInt(element.dataset.itemId, 10);
        this.name = element.dataset.itemName;
        this.departmentId = parseInt(element.dataset.departmentId, 10);
        this.qty = this._extractQtyFromInputElement();

        return this;
    }

    _extractQtyFromInputElement() {
        const qtyElement = document.getElementById(`item-${this.itemId}-qty`);

        return parseInt(qtyElement.value, 10);
    }
}
