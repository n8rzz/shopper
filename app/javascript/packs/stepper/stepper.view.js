const CLASSNAMES = {
    ROOT: 'js-stepper',
    DECREASE: 'js-stepper-decrease',
    VALUE: 'js-stepper-value',
    INCREASE: 'js-stepper-increase',
    SUBMIT: 'js-stepper-submit',
};

const ACTION = {
    DECREASE: 'DECREASE',
    INCREASE: 'INCREASE',
};

/**
 * Connects to a number input with external decrease/increase
 * action elements in addition to a submit button
 *
 * accepts `onSubmitHandler` from parent
 *
 * This class is concerned with decreasin/increasing a value of
 * a number field, and submitting that current value onClick
 *
 * @class StepperView
 */
export default class StepperView {
    constructor($element, onSubmitHandler) {
        this._id = -1;
        this._$element = $element;
        this._onSubmitHandler = onSubmitHandler;

        if (!this._$element) {
            return null;
        }

        return this._createChildren()
            ._setupHandlers()
            .enable();
    }

    _createChildren() {
        this._id = parseInt(this._$element.dataset.stepperId, 10);
        this._$decreaseButton = document.getElementsByClassName(this._buildClassnameWithId(CLASSNAMES.DECREASE))[0];
        this._$increaseButton = document.getElementsByClassName(this._buildClassnameWithId(CLASSNAMES.INCREASE))[0];
        this._$valueView = document.getElementsByClassName(this._buildClassnameWithId(CLASSNAMES.VALUE))[0];
        this._$submitButton = document.getElementsByClassName(this._buildClassnameWithId(CLASSNAMES.SUBMIT))[0];

        return this;
    }

    _setupHandlers() {
        this._onClickDecreaseHandler = this._onClickDecrease.bind(this);
        this._onClickIncreaseHandler = this._onClickIncrease.bind(this);
        this._onClickSubmitHandler = this._onClickSubmit.bind(this);

        return this;
    }

    enable() {
        this._$decreaseButton.addEventListener('click', this._onClickDecreaseHandler);
        this._$increaseButton.addEventListener('click', this._onClickIncreaseHandler);
        this._$submitButton.addEventListener('click', this._onClickSubmitHandler);

        return this;
    }

    _buildClassnameWithId(classname) {
        return `${classname}-${this._id}`;
    }

    _onClickDecrease(event) {
        this._updateValueView(ACTION.DECREASE);
    }

    _onClickIncrease(event) {
        this._updateValueView(ACTION.INCREASE);
    }

    _onClickSubmit(event) {
        const currentValue = parseInt(this._$valueView.value, 10);

        this._onSubmitHandler(event, currentValue);
        this._resetValue();
    }

    _resetValue() {
        this._$valueView.value = 1;
    }

    _updateValueView(action) {
        const currentValue = parseInt(this._$valueView.value, 10);
        let nextValue;

        switch(action) {
            case ACTION.DECREASE:
                if (currentValue <= 1) {
                    return;
                }

                nextValue = currentValue - 1;

                break;
            case ACTION.INCREASE:
                nextValue = currentValue + 1;

                break;
            default:
                break;
        }

        this._$valueView.value = nextValue;
    }
}
