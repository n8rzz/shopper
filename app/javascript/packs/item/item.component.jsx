import React from 'react';
import PropTypes from 'prop-types';
import { StepperContainer } from '../components/stepper/stepper.container';

export function Item(props) {
    const editItemPath = `items/${props.itemId}/edit`;

    return (
        <li className={'js-item'}>
            <div className={'listItem mix-listItem_flex'}>
                <div className={'listItem-content'}>
                    <div className={'listItem-hd'}>
                        <h2 className={'hdg hdg_2'}>
                            <a href={editItemPath}>
                                {props.itemName}
                            </a>
                        </h2>
                    </div>
                    <div className={'listItem-ft'}>
                        { props.departmentName }
                    </div>
                </div>
                <div className={'listItem-action'}>
                    <StepperContainer
                        csrf={props.csrf}
                        departmentId={props.departmentId}
                        itemId={props.itemId}
                    />
                </div>
            </div>
        </li>
    );
}

Item.propTypes = {
    csrf: PropTypes.string.isRequired,
    departmentId: PropTypes.number.isRequired,
    departmentName: PropTypes.string.isRequired,
    itemId: PropTypes.number.isRequired,
    itemName: PropTypes.string.isRequired,
};
