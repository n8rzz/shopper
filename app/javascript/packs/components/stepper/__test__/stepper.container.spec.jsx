import React from 'react';
import sinon from 'sinon';
import { mount, shallow } from 'enzyme';

import { StepperContainer } from '../stepper.container';
import { Stepper } from '../stepper';
import { csrfMock } from '../../../__mocks__/csrf.mock';
import { itemMock } from '../../../__mocks__/item.mock';

describe('StepperContainer', () => {
    let component = null;

    beforeEach(() => {
        component = mount(<StepperContainer
            csrf={csrfMock}
            departmentId={itemMock.department_id}
            itemId={itemMock.id}
            itemName={itemMock.name}
            isSubmitting={false}
            isSubmitSuccess={false}
            onSubmitHandler={sinon.spy()}
        />);
    });

    afterEach(() => {
        component = null;
    });

    test('renders correctly with valid props', () => {
        const shallowComponent = shallow(<StepperContainer
            csrf={csrfMock}
            departmentId={itemMock.department_id}
            itemId={itemMock.id}
            itemName={itemMock.name}
            isSubmitting={false}
            isSubmitSuccess={false}
            onSubmitHandler={sinon.spy()}
        />);

        expect(shallowComponent).toMatchSnapshot();
    });

    describe('decrease btn element', () => {
        test('contains decrease btn element', () => {
            const decreaseBtn = component.find('.js-stepper-btn-decrease');

            expect(decreaseBtn.length).toBe(1);
        });

        test('responds to click', () => {
            const onClickDecreaseHandlerSpy = sinon.spy();

            component.props.onClickDecreaseHandler = onClickDecreaseHandlerSpy;
            const stepperChild = component.find(Stepper);

            stepperChild.props.onClickDecreaseHandler();

            expect(onClickDecreaseHandlerSpy.callCount).toBe(1);
        });
    });

    describe('increase btn element', () => {
        test('contains increase btn element', () => {
            const increaseBtn = component.find('.js-stepper-btn-increase');

            expect(increaseBtn.length).toBe(1);
        });

        test('responds to click', () => {
            const onClickIncreaseHandlerSpy = sinon.spy();

            component.props.onClickIncreaseHandler = onClickIncreaseHandlerSpy;
            const stepperChild = component.find(Stepper);

            stepperChild.props.onClickIncreaseHandler();

            expect(onClickIncreaseHandlerSpy.callCount).toBe(1);
        });
    });
});
