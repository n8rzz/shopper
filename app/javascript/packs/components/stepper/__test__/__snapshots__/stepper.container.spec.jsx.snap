// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StepperContainer renders correctly with valid props 1`] = `
<Stepper
  isSubmitSuccess={false}
  isSubmitting={false}
  itemId={15}
  onChangeQtyHandler={[Function]}
  onClickDecreaseHandler={[Function]}
  onClickIncreaseHandler={[Function]}
  onClickSubmitHandler={[Function]}
  qty={1}
/>
`;
